docs/architecture/adr-038-state-listening.md

```diff
@@ -372,59 +571,122 @@ func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx
+system over gRPC that is used to load and run Cosmos-SDK plugins. The plugin system uses [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin).
+Each plugin must have a struct that implements the `plugin.Plugin` interface and an `Impl` interface for processing messages over gRPC.
+Each plugin must also have a message protocol defined for the gRPC service:
 
 ```go
-// Plugin is the base interface for all kinds of cosmos-sdk plugins
+// ListenerPlugin is the base struc for all kinds of go-plugin implementations
 // It will be included in interfaces of different Plugins
-type Plugin interface {
-	// Name should return unique name of the plugin
-	Name() string
-
-	// Version returns current version of the plugin
-	Version() string
+type ListenerPlugin interface {
+	// GRPCPlugin must still implement the Plugin interface
+	plugin.Plugin
+	// Concrete implementation, written in Go. This is only used for plugins
+	// that are written in Go.
+	Impl baseapp.ABCIListener
+}
 
-	// Init is called once when the Plugin is being loaded
-	// The plugin is passed the AppOptions for configuration
-	// A plugin will not necessarily have a functional Init
-	Init(env serverTypes.AppOptions) error
+func (p *ListenerGRPCPlugin) GRPCServer(_ *plugin.GRPCBroker, s *grpc.Server) error {
+	RegisterABCIListenerServiceServer(s, &GRPCServer{Impl: p.Impl})
+	return nil
+}
 
-	// Closer interface for shutting down the plugin process
-	io.Closer
+func (p *ListenerGRPCPlugin) GRPCClient(
+	_ context.Context,
+	_ *plugin.GRPCBroker,
+	c *grpc.ClientConn,
+) (interface{}, error) {
+	return &GRPCClient{client: NewABCIListenerServiceClient(c)}, nil
+>>>>>>> Stashed changes
 }
 ```
 
-The `Name` method returns a plugin's name.
-The `Version` method returns a plugin's version.
-The `Init` method initializes a plugin with the provided `AppOptions`.
-The io.Closer is used to shut down the plugin service.
+The `plugin.Plugin` interface has two methods `Client` and `Server`. For our GRPC service these are `GRPCClient` and `GRPCServer`
+The `Impl` field holds the concrete implementation of our `baseapp.ABCIListener` interface written in Go.
+Note: this is only used for plugins that are written in Go.
+
+The `StreamingService` protocol:
+
+```protobuf
+syntax = "proto3";
+
+package cosmos.sdk.grpc.abci.v1;
+
+option go_package           = "github.com/cosmos/cosmos-sdk/streaming/plugins/abci/grpc_abci_v1";
+option java_multiple_files  = true;
+option java_outer_classname = "AbciListenerProto";
+option java_package         = "network.cosmos.sdk.grpc.abci.v1";
+
+// PutRequest is used for storing ABCI request and response
+// and Store KV data for streaming to external grpc service.
+message PutRequest {
+  int64 block_height      = 1;
+  bytes req               = 2;
+  bytes res               = 3;
+  bytes store_kv_pair     = 4;
+  int64 store_kv_pair_idx = 5;
+  int64 tx_idx            = 6;
+}
+
+message Empty {}
+
+service ABCIListenerService {
```

Comment:
nit: might only be me but I find the notion of ABCI here slightly confusing. anyway to name it something else

Reply:
we can ignore for now and come back to alter